### ISSUE TITLE
Cannot solve specific dns record on kubernetes (coredns / nodelocaldns)

### ISSUE THREAD

**Issue by lderugeriis** [miekg/coredns#6346]
Hi everyone,
I hit a weird problem with coredns/nodelocaldns, basically I can't resolve a specific dns record for a domain. The record is `www.digicert.com` , it answers correctly tho if I dig with *+tcp* or use *force_tcp* in my nodelocaldns Corefile or use 8.8.8.8 as a forwarder.
This is my Corefile.
```
    .:53 {
        errors
        cache 30
        reload
        loop
        bind 169.254.25.10
        forward . /etc/resolv.conf
        prometheus :9253
    }
```
If I use udp all I'm getting is a timeout error
```
[ERROR] plugin/errors: 2 www.digicert.com. A: read udp 169.254.25.10:40601->169.254.25.10:53: i/o timeout
[ERROR] plugin/errors: 2 www.digicert.com. A: read udp k8sworkernode:59755->mycompanyns1:53: i/o timeout
[ERROR] plugin/errors: 2 www.digicert.com. A: read udp 169.254.25.10:52648->169.254.25.10:53: i/o timeout
[ERROR] plugin/errors: 2 www.digicert.com. A: read udp k8sworkernode:42793->mycompanyns2:53: i/o timeout
```
All the other requests work correctly (dig google.com, even dig digicert.com etc.), and I also have deployed two more (staging) clusters on other subnets and the name get resolved without problems.
I checked udp ports and they are open, I rebooted kubernetes nodes...
Below is a tcpdump on a staging cluster where  resolution works (www.digicert.com is a CNAME for 3ivgup7.x.incapdns.net)
```
19:31:42.472765 IP k8snode.55631 > ns2.mycompany: 8824+ [1au] A? 3ivgup7.x.incapdns.net. (63)
19:31:42.474636 IP ns2.mycompany > k8snode.55631: 8824 1/6/43 A 45.60.123.229 (847)
19:31:44.195436 IP k8snode.55631 > ns2.mycompany: 26109+ AAAA? memcache. (26)
19:31:44.195721 IP k8snode.39894 > ns1.mycompany: 44597+ A? memcache. (26)
19:31:44.196073 IP ns2.mycompany > k8snode.55631: 26109 NXDomain 0/1/0 (101)
19:31:44.196507 IP ns1.mycompany > k8snode.39894: 44597 NXDomain 0/1/0 (101)
```
and here on a production cluster where it doesn't
```
19:28:15.860733 IP k8snode.54014 > ns2.mycompany: 10512+ [1au] A? 3ivgup7.x.incapdns.net. (63)
19:28:17.861279 IP k8snode.35119 > ns1.mycompany: 6410+ [1au] A? 3ivgup7.x.incapdns.net. (63)
19:28:17.861307 IP k8snode.59769 > ns2.mycompany: 1440+ NS? . (17)
19:28:17.862230 IP ns2.mycompany > k8snode.59769: 1440 13/0/13 NS i.root-servers.net., NS c.root-servers.net., NS f.root-servers.net., NS b.root-servers.net., NS a.root-servers.net., NS d.root-servers.net., NS j.root-servers.net., NS k.root-servers.net., NS m.root-servers.net., NS l.root-servers.net., NS e.root-servers.net., NS h.root-servers.net., NS g.root-servers.net. (508)
19:28:19.862581 IP k8snode.47932 > ns1.mycompany: 20584+ NS? . (17)
19:28:19.862813 IP k8snode.57677 > ns2.mycompany: 31162+ [1au] A? 3ivgup7.x.incapdns.net. (63)
19:28:19.863374 IP ns1.mycompany > k8snode.47932: 20584 13/0/13 NS l.root-servers.net., NS a.root-servers.net., NS i.root-servers.net., NS h.root-servers.net., NS m.root-servers.net., NS g.root-servers.net., NS d.root-servers.net., NS c.root-servers.net., NS e.root-servers.net., NS f.root-servers.net., NS j.root-servers.net., NS b.root-servers.net., NS k.root-servers.net. (508)
19:28:20.860937 IP k8snode.56137 > ns2.mycompany: 24068+ [1au] A? 3ivgup7.x.incapdns.net. (63)
19:28:21.862959 IP k8snode.39945 > ns1.mycompany: 32140+ [1au] A? 3ivgup7.x.incapdns.net. (63)
19:28:21.863103 IP k8snode.46998 > ns2.mycompany: 40873+ NS? . (17)
```
And it repeats endlessly
Both cluster have the same /etc/resolv.conf 
```
domain cluster.local
search default.svc.cluster.local svc.cluster.local
nameserver 169.254.25.10
nameserver companyns1
nameserver companyns2
options ndots:2 timeout:2 attempts:2
```
and identical coredns configuration and version. I installed kubernetes with kubespray and I'm using antrea if that matters, on Debian 10.
I'm a bit lost at the moment...

**Comment by lderugeriis** [comment]
Turned out (after much much troubleshooting) that our internal NS, which runs on bind, answered with additional entries that caused the packet to be larger than 4096 bites. Bind default size for udp packet is 4096 which is larger than the default MTU so the packet get fragmented. That way the fragmented packets don't match the "permit" rule on our router ACLs for port 53 (because the source port is missing from the packet) so it couldn't send the answer back to kubernetes nodes causing a timeout on coredns.
We added 
```
        max-udp-size 512;
        edns-udp-size 512;
        minimal-responses yes;
```
on bind configuration in order for the query to be switched in tcp if larger than 512 bytes and that solved the issue.
The fact that I'd like still to comprehend is, why I could reproduce the problem only if I dig with +dnssec? Because if i dig without any parametres using our DNS as they were configured, they answered correctly. 
Does coredns verify dnssec records by default?
